Move Custom Env from tab to grouped select in Application Compute Configuration

ehanna4 opened the pull request:
The majority of the changes associated with [SATURN-1261]
note the changes in common.js are being reviewed in another PR so no need to review that here I will merge that into here once review is complete

To do which will be seen in another PR:
 - move trash can out of widget and into drawer (and add the delete warning view)
 - move the back arrow to the right side of the warning views

Tested:
- tested by observing changes in the browser - all previous behavior is the same just the look of the drawer is changed: tab is gone and the select dropdown is grouped
- created notebooks (both custom and jupyter and observed that the drawer continues to open to the currently created cluster as is expected)
- tested out the regular cluster options, starting, stopping, deleting etc. and all behaved as expected
 

[SATURN-1261]: https://broadworkbench.atlassian.net/browse/SATURN-1261

ehanna4 committed:
Application Compute Config Drawer Refactor

- Change name to Application Compute
- Update Select to be able to have grouped sections
- Remove tab bar from the configuration drawer

TODO:
- match style of groups to mocks
- move delete option inside drawer
- ensure current cluster sets the default

ehanna4 committed:
Use flatmap over flatten and map

ehanna4 committed:
Update selector per PR reviews

ehanna4 commented on src/components/NewClusterModal.js:
```diff
@@ -111,7 +111,7 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
       profile: matchingProfile ? matchingProfile.name : 'custom',
       jupyterUserScriptUri: '',
-      isCustomEnv: false, customEnvImage: '', viewMode: undefined,
+      customEnvImage: '', viewMode: 'JupyterEnv', prevViewMode: 'JupyterEnv',
```
`prevViewMode` is more important when the `delete warn view` is added to the drawer since you will be able to delete from any of the environment views (Jupyter, Custom, and Other) having the `prevViewMode` ensures that clicking the `back arrow` or `back button` will return the user to the correct view.

zarsky-broad replied:
Instead of this, why not build an array of where you are and have been, breadcrumb style?

ehanna4 replied:
I hadn't considered that, but it sounds like a more straightforward approach, I will look into that. Thanks!

ehanna4 committed:
Show correct currentcluster

- drawer opens to current cluster configs

- todo: if you put the image url of a jupyter image in the custom env it switches to the corresponding jupyter select value is this an issue?

ehanna4 committed:
Update select and use breadcrumbs for tracking drawer view

ehanna4 committed:
Merge branch 'dev' into eh-r-studio-widget-update

ehanna4 committed:
Clean up code

To do:
- reduce diff clutter
- replace unnecessary grids with flexes
- add delete boolean and make switch a cond

ehanna4 committed:
Replace unnecessary grid styling

ehanna4 commented:
Ready for review. The final pieces of this ticket (changing the back arrow position, moving the delete into the drawer, and having a delete warning page) will be followed up in subsequent PRs.

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -216,209 +375,77 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
-              h(Fragment, [
+          div({ style: { marginBottom: '1rem' } },
+            ['Create a compute instance to launch Jupyter Notebooks or a Project-Specific software application.']),
+          div({ style: { minHeight: 100 } },
```
nit: This `minHeight` doesn't seem to be doing anything any more, so we can probably get rid of this div entirely.

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -216,209 +375,77 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+            ['Create a compute instance to launch Jupyter Notebooks or a Project-Specific software application.']),
+          div({ style: { minHeight: 100 } },
+            [
+              div([
```
nit: This div can be removed.

breilly2 commented on src/components/NewClusterModal.js:
```diff
@@ -185,22 +185,181 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+        ])
+    ])
+
+    const machineConfig = () => h(Fragment, [
```
I assume this is an unmodified extracted function. Let me know if there are any notable changes to review.

ehanna4 replied:
correct - machineConfig did not get changed at all, just moved

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -185,22 +185,181 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+          div([
+            h(ButtonPrimary, {
+              disabled: selectedLeoImage === CUSTOM_MODE && isCustomImageInvalid,
+              tooltip: selectedLeoImage === CUSTOM_MODE && isCustomImageInvalid &&
```
It might be nice to extract `selectedLeoImage === CUSTOM_MODE` to a local variable, since it's repeated several times throughout.

breilly2 commented on src/components/common.js:
```diff
@@ -304,9 +304,7 @@ const commonSelectProps = {
-export const Select = ({ value, options, id, ...props }) => {
-  const newOptions = options && !_.isObject(options[0]) ? _.map(value => ({ value }), options) : options
-  const findValue = target => _.find({ value: target }, newOptions)
+const BaseSelect = ({ value, newOptions, id, findValue, ...props }) => {
```
Please update docs for this function and/or provide docs for the exported `Select` and `GroupedSelect` functions.

ehanna4 replied:
good catch!

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -216,209 +375,77 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
-                      ]
-                    })
+                    div({ style: { marginBottom: '0.5rem' } }, [label({ htmlFor: id, style: styles.label }, 'ENVIRONMENT')]),
+                    div({ style: { height: '45px' } }, [makeGroupedEnvSelect(id)])
```
The validation error ends up a little cramped with this at 45px. We should bump it up to at least 52px or so.

panentheos reviewed (approved):
A couple minor markup suggestions, but otherwise looks good.

Tested by creating clusters from both standard and custom images.

breilly2 reviewed (approved):
One request about function docs then 👍. Looks good! I tested lots of interactions in the application compute widget, including replacing a pre-installed runtime with a custom runtime (however, I'm currently seeing a problem because the image list we're getting has `image` instead of `imageUrl` properties... but this appears to be a change on the leo side, not our side).

breilly2 commented:
Update on issue I was seeing: it was because I was using an RStudio custom image which we won't recognize until #1977 lands.

ehanna4 committed:
Peer Review feedback

ehanna4 committed:
Merge branch 'dev' into eh-r-studio-widget-update
